### PR TITLE
[SYCLomatic] Support DeclRefExpr in dim info analysis and move the analysis logic to common code path

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 1]
```diff
@@ -1213,6 +1215,138 @@ int KernelCallExpr::calculateOriginArgsSize() const {
...
+    return 3;
+
+  // VD's DRE should be only used once (as the config arg) in VDContext
+  auto DREMatcher =
+      ast_matchers::findAll(ast_matchers::declRefExpr().bind("DRE"));
+  auto MatchedResults =
+      ast_matchers::match(DREMatcher, *VDContext, DpctGlobalInfo::getContext());
+  size_t RefCount = 0;
+  for (const auto &Res : MatchedResults) {
+    const DeclRefExpr *RefExpr = Res.getNodeAs<DeclRefExpr>("DRE");
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]: has_decl(VD)

Comment (zhiweij1) [reply clang/lib/DPCT/AnalysisInfo.cpp]: done

**Review by tomflinda — APPROVED** [review]
LGTM

File: clang/include/clang/ASTMatchers/ASTMatchers.h [hunk 2]
```diff
@@ -8288,6 +8288,13 @@ AST_MATCHER_P(Expr, ignoringElidableConstructorCall,
   return InnerMatcher.matches(Node, Finder, Builder);
 }
 
+#ifdef SYCLomatic_CUSTOMIZATION
```
Comment (zhiweij1) [inline clang/include/clang/ASTMatchers/ASTMatchers.h]: move to dpct code file

File: clang/include/clang/ASTMatchers/ASTMatchers.h [hunk 3]
```diff
@@ -8288,6 +8288,13 @@ AST_MATCHER_P(Expr, ignoringElidableConstructorCall,
...
+AST_MATCHER_P(VarDecl, isSameAs, const VarDecl *, TargetVD) {
```
Comment (zhiweij1) [inline clang/include/clang/ASTMatchers/ASTMatchers.h]:
```suggestion
AST_MATCHER_P(DeclRefExpr, isSameAs, const VarDecl *, TargetVD) {
```
Move getdecl and check here

File: clang/lib/DPCT/ExprAnalysis.cpp [hunk 4]
```diff
@@ -2004,6 +2000,51 @@ void KernelConfigAnalysis::analyzeExpr(const CXXTemporaryObjectExpr *Ctor) {
...
+      ast_matchers::match(DREMatcher, *VDContext, DpctGlobalInfo::getContext());
+  if (MatchedResults.size() != 1)
+    return ArgumentAnalysis::analyzeExpr(DRE);
+
+  if (VD->getType().getCanonicalType().getAsString() != "struct dim3")
+    return ArgumentAnalysis::analyzeExpr(DRE);
+  if (!VD->hasInit())
+    return ArgumentAnalysis::analyzeExpr(DRE);
+
+  const CXXConstructExpr *Init = dyn_cast<CXXConstructExpr>(VD->getInit());
```
Comment (zhiweij1) [inline clang/lib/DPCT/ExprAnalysis.cpp]:
```suggestion
  if (VD->getInit())
    dispatch(VD->getInit());
  else
    return ArgumentAnalysis::analyzeExpr(DRE);
```

File: clang/lib/DPCT/ExprAnalysis.cpp [hunk 5]
```diff
@@ -2004,6 +2000,51 @@ void KernelConfigAnalysis::analyzeExpr(const CXXTemporaryObjectExpr *Ctor) {
...
+  if (!Init)
+    return ArgumentAnalysis::analyzeExpr(DRE);
+
+  if (IsTryToUseOneDimension) {
+    // Insert member access expr at the end of DRE
+    addReplacement(getExprLength(), 0, ".get(2)");
+  }
+
+  handleDim3Ctor(Init, SourceRange(Init->getBeginLoc(), Init->getEndLoc()),
```
Comment (zhiweij1) [inline clang/lib/DPCT/ExprAnalysis.cpp]:
```suggestion

```

File: clang/lib/DPCT/ExprAnalysis.cpp [hunk 6]
```diff
@@ -2004,6 +2000,51 @@ void KernelConfigAnalysis::analyzeExpr(const CXXTemporaryObjectExpr *Ctor) {
...
+  // KernelCallExpr
+  if (!VD->isLocalVarDecl())
+    return ArgumentAnalysis::analyzeExpr(DRE);
+  const CompoundStmt *VDContext =
+      DpctGlobalInfo::findAncestor<CompoundStmt>(VD);
+  const CompoundStmt *KernelCallContext =
+      DpctGlobalInfo::findAncestor<CompoundStmt>(getTargetExpr());
+  if (!VDContext || !KernelCallContext || (VDContext != KernelCallContext))
+    return ArgumentAnalysis::analyzeExpr(DRE);
+
```
Comment (zhiweij1) [inline clang/lib/DPCT/ExprAnalysis.cpp:2020]:
```suggestion
using namespace ast_matchers
```

File: clang/lib/DPCT/ExprAnalysis.cpp [hunk 7]
```diff
@@ -2004,6 +2000,51 @@ void KernelConfigAnalysis::analyzeExpr(const CXXTemporaryObjectExpr *Ctor) {
...
 }
 
+void KernelConfigAnalysis::analyzeExpr(const DeclRefExpr *DRE) {
+  const VarDecl *VD = dyn_cast_or_null<VarDecl>(DRE->getDecl());
+  if (!VD)
+    return ArgumentAnalysis::analyzeExpr(DRE);
+
+  // VD must be local variable and must have the same context with
+  // KernelCallExpr
+  if (!VD->isLocalVarDecl())
```
Comment (zhiweij1) [inline clang/lib/DPCT/ExprAnalysis.cpp:2010]: dyn_cast<FunctionDecl>(VD->getParentContext())

Comment (zhiweij1) [inline clang/lib/DPCT/ExprAnalysis.cpp:2010]:
```suggestion
dyn_cast<FunctionDecl>(VD->getParentContext())
```

File: clang/lib/DPCT/ExprAnalysis.cpp [hunk 8]
```diff
@@ -2004,6 +2000,51 @@ void KernelConfigAnalysis::analyzeExpr(const CXXTemporaryObjectExpr *Ctor) {
...
+    return ArgumentAnalysis::analyzeExpr(DRE);
+  const CompoundStmt *VDContext =
```
Comment (zhiweij1) [inline clang/lib/DPCT/ExprAnalysis.cpp]: remove 2012 ~ 2017 , use the impl of isLocalVarDecl

File: clang/lib/DPCT/Utility.cpp [hunk 9]
```diff
@@ -1972,10 +1972,12 @@ SourceLocation getEndLocOfFollowingEmptyMacro(SourceLocation Loc) {
   auto &SM = dpct::DpctGlobalInfo::getSourceManager();
   auto &Map = dpct::DpctGlobalInfo::getBeginOfEmptyMacros();
   Token Tok;
-  Lexer::getRawToken(
+  bool Ret = Lexer::getRawToken(
       Loc.getLocWithOffset(Lexer::MeasureTokenLength(
           Loc, SM, dpct::DpctGlobalInfo::getContext().getLangOpts())),
       Tok, SM, dpct::DpctGlobalInfo::getContext().getLangOpts(), true);
+  if (Ret)
+    return Loc;
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp:1980]: Confirm with Andy

File: clang/include/clang/ASTMatchers/ASTMatchers.h (hunk 2 above)
Comment (zhiweij1) [reply clang/include/clang/ASTMatchers/ASTMatchers.h]: done

File: clang/include/clang/ASTMatchers/ASTMatchers.h (hunk 3 above)
Comment (zhiweij1) [reply clang/include/clang/ASTMatchers/ASTMatchers.h]: done

File: clang/lib/DPCT/ExprAnalysis.cpp (hunk 6 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ExprAnalysis.cpp:2020]: done

File: clang/lib/DPCT/ExprAnalysis.cpp (hunk 5 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ExprAnalysis.cpp]: done

File: clang/lib/DPCT/ExprAnalysis.cpp (hunk 4 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ExprAnalysis.cpp]: done

File: clang/lib/DPCT/ExprAnalysis.cpp (hunk 8 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ExprAnalysis.cpp]: done

File: clang/include/clang/ASTMatchers/ASTMatchers.h (hunk 2 above)
Comment (zhiweij1) [reply clang/include/clang/ASTMatchers/ASTMatchers.h]: done

File: clang/lib/DPCT/Utility.cpp (hunk 9 above)
Comment (zhiweij1) [reply clang/lib/DPCT/Utility.cpp:1980]: done

File: clang/lib/DPCT/ExprAnalysis.cpp [hunk 10]
```diff
@@ -2004,6 +2000,44 @@ void KernelConfigAnalysis::analyzeExpr(const CXXTemporaryObjectExpr *Ctor) {
...
+  const Stmt *VDContext = FD->getBody();
+
+  // VD's DRE should be only used once (as the config arg) in VDContext
+  auto DREMatcher = findAll(declRefExpr(isDeclSameAs(VD)).bind("DRE"));
+  auto MatchedResults =
+      match(DREMatcher, *VDContext, DpctGlobalInfo::getContext());
+  if (MatchedResults.size() != 1)
+    return ArgumentAnalysis::analyzeExpr(DRE);
+
+  if (VD->getType().getCanonicalType().getAsString() != "struct dim3")
```
Comment (zhiweij1) [inline clang/lib/DPCT/ExprAnalysis.cpp]: can remove

File: clang/lib/DPCT/ExprAnalysis.cpp [hunk 11]
```diff
@@ -2004,6 +2000,44 @@ void KernelConfigAnalysis::analyzeExpr(const CXXTemporaryObjectExpr *Ctor) {
...
+    return ArgumentAnalysis::analyzeExpr(DRE);
+  if (!VD->hasInit())
+    return ArgumentAnalysis::analyzeExpr(DRE);
+
+  if (!VD->getInit())
```
Comment (zhiweij1) [inline clang/lib/DPCT/ExprAnalysis.cpp]: remove

File: clang/lib/DPCT/ExprAnalysis.cpp [hunk 12]
```diff
@@ -2004,6 +2000,44 @@ void KernelConfigAnalysis::analyzeExpr(const CXXTemporaryObjectExpr *Ctor) {
   return ArgumentAnalysis::analyzeExpr(Ctor);
 }
 
+void KernelConfigAnalysis::analyzeExpr(const DeclRefExpr *DRE) {
+  using namespace ast_matchers;
```
Comment (zhiweij1) [inline clang/lib/DPCT/ExprAnalysis.cpp:2007]: return if not 1st or 2nd argument

File: clang/lib/DPCT/ExprAnalysis.cpp (hunk 10 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ExprAnalysis.cpp]: done

File: clang/lib/DPCT/ExprAnalysis.cpp (hunk 11 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ExprAnalysis.cpp]: done

File: clang/lib/DPCT/ExprAnalysis.cpp (hunk 12 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ExprAnalysis.cpp:2007]: done

**Review by ziranzha — APPROVED** [review]

**Pull request merged**